### DISCUSSION TITLE
Issue with ls running ahead, and trying to use a random file

### DISCUSSION THREAD

**Discussion by tdardly** [savonet/liquidsoap#3727] [User Support]
I've been trying to stop liquidsoap going crazy, and running ahead with ffmpeg trying to read all the files (of many thousands) all at once... well, in very quick succession. I don't understand why it's trying to do that, but my CPU just maxes out so I've been trying to get ls instead just to pick a random file to play, when it needs to. 

So I've re-written this so many times, but this is the current script. Many (many) arguments with ChatGPT who insists this should definitely work, but ls complains that the line:

```
random_track = list.nth(playlist_lines, random(list.length(playlist_lines)))
```

is returning an integer... (char 17-31)

```
Error 5: this value has type
  int
but it should be a subtype of the type of the value at switches.liq, line 120, char 24-31
  [_]
```

Would so appreciate someone telling me where I'm going wrong. I've spent days on this, and I've almost run out of time.

```
# Configuration Variables
MUSIC_DIR = "/data/music"
PROMO_DIR = "/data/promo"
MUSIC_PLAYLIST = "#{MUSIC_DIR}/pl.m3u"
LOGFILE = "/data/logs/liquidsoap.log"
ICECAST_HOST = "my.icecast.server"
ICECAST_PORT = 80
ICECAST_PASS = "hackme"
MUSIC_MOUNT = "/128.mp3"
REBROADCAST_MOUNT = "/64.aac"
REBROADCAST_CODEC = %fdkaac(aot="mpeg4_aac_lc", bitrate=64)
MP3_CODEC = %mp3(bitrate=128)
INPUT_STREAM_URL = "http://" ^ ICECAST_HOST ^ "/128.mp3"

# Function to dynamically load a track from the global playlist
def dynamic_music_request() =
  # Get the lines from the playlist
  playlist_lines = get_process_lines("cat #{MUSIC_PLAYLIST}")
  
  random_track = list.nth(playlist_lines, random(list.length(playlist_lines)))
  request.create("#{MUSIC_DIR}/" ^ random_track)
end

# Create a dynamic source for music
dynamic_music = request.dynamic(id="dynamic_music", dynamic_music_request)

# Make the music source safe
safe_music = mksafe(dynamic_music)

# Function to dynamically load a promo
def dynamic_promo() =
  request.create("#{PROMO_DIR}/" ^ list.hd(get_process_lines("ls -1 #{PROMO_DIR}")))
end

# Create dynamic source for promos
promos = mksafe(request.dynamic(dynamic_promo))

# Weight tracks and promos (2:1 ratio)
weighted = random(weights=[2, 1], [safe_music, promos])
safe_weighted = mksafe(weighted)

# Crossfade between tracks
crossfaded = mksafe(crossfade(safe_weighted))

# Output to Icecast as MP3
output.icecast(%mp3,
  host = ICECAST_HOST,
  port = ICECAST_PORT,
  password = ICECAST_PASS,
  mount = MUSIC_MOUNT,
  crossfaded)

# Input stream (re-encode and rebroadcast)
input_stream = mksafe(input.http(INPUT_STREAM_URL))

# Re-encode and output to Icecast
output.icecast(REBROADCAST_CODEC,
  host = ICECAST_HOST,
  port = ICECAST_PORT,
  password = ICECAST_PASS,
  mount = REBROADCAST_MOUNT,
  input_stream)




```

**Comment by toots** [comment]
Thanks for filling this.

ChatGPT is a nice assistant but it's still giving you what's most statistically likely. It doesn't read our doc and I strongly suggest the answer to your question is there :-)

The doc for the operator you are considering is here: https://www.liquidsoap.info/doc-2.2.4/reference.html#random

As you can see, the operator is of type:
```liquidsoap
(?id : string?, ?override : string, ?replay_metadata : bool,
 ?transition_length : float,
 ?transitions : [(source('a), source('a)) -> source('a)], ?weights : 
 [{int}], [source('a)]) -> source('a)
```

i.e. it returns a `source`! Not what you want!

The operator you want is here: https://www.liquidsoap.info/doc-2.2.4/reference.html#random.int

> random.int
> Generate a random value between min (included) and max (excluded).
> 
> Type:
> 
> (?min : int, ?max : int) -> int
> Arguments:
> 
> min (of type int, which defaults to -536870911)
> max (of type int, which defaults to 536870912)

You should use it like this:
```liquidsoap
random_track = list.nth(playlist_lines, random.int(min=0, max=list.length(playlist_lines)))
```

**Reply by tdardly** [reply]
That works, thanks! But, why is it still processing another file every ~10s? I'm guessing this is why it's still maxing out the CPU Icecast is eventually kicking off clients for 'falling behind'?

**Reply by tdardly** [reply]
Is this anything to do with "clocks" I wonder? I should probably know this, but I'm not totally sure...

**Reply by tdardly** [reply]
Also, if relevant: 
```
[mp3float @ 0x7fbe8d981400] Could not update timestamps for discarded samples.
```

**Comment by tdardly** [comment]
Ok - I've worked out what it is. It's the second input/output. This:
```
input_stream = mksafe(input.http(INPUT_STREAM_URL))

# Re-encode and output to Icecast
output.icecast(REBROADCAST_CODEC,
  host = ICECAST_HOST,
  port = ICECAST_PORT,
  password = ICECAST_PASS,
  mount = REBROADCAST_MOUNT,
  input_stream)
```
Somehow, it's interfering with the autodj and convincing it that it needs to read more and more tracks. If I run that bit of code in a separate docker container, it's all fine.

I'd thought that it was ok to have multiple inputs/outputs - so is there a way to make this code co-exist, perhaps different naming or something? Or do I need to start another process so it's completely separate?

**Reply by toots** [reply]
Yeah you don't need to re-input the stream, you can just plug a second output out of it:

```
# Re-encode and output to Icecast
output.icecast(REBROADCAST_CODEC,
  host = ICECAST_HOST,
  port = ICECAST_PORT,
  password = ICECAST_PASS,
  mount = REBROADCAST_MOUNT,
  crossfaded)
```

**Reply by tdardly** [reply]
Ah unfortunately there's another stream that hits the 'main' mountpoint. autodj is streaming to the backup mount. So that bit of code needs to reencode whatever is on the main stream, whether it's the external stream, or the autodj stuff from the fallback...

**Comment by vitoyucepi** [comment]
Hi @tdardly,
```ruby
# Function to dynamically load a track from the global playlist
def dynamic_music_request() =
  # Get the lines from the playlist
  playlist_lines = get_process_lines("cat #{MUSIC_PLAYLIST}")
  
  random_track = list.nth(playlist_lines, random(list.length(playlist_lines)))
  request.create("#{MUSIC_DIR}/" ^ random_track)
end

# Create a dynamic source for music
dynamic_music = request.dynamic(id="dynamic_music", dynamic_music_request)

# Function to dynamically load a promo
def dynamic_promo() =
  request.create("#{PROMO_DIR}/" ^ list.hd(get_process_lines("ls -1 #{PROMO_DIR}")))
end

# Create dynamic source for promos
promos = mksafe(request.dynamic(dynamic_promo))
```

Do you *really* need `request.dynamic` here? Could you explain what you are trying to accomplish?
Perhaps a [`playlist`](https://www.liquidsoap.info/doc-2.2.4/reference.html#playlist) would be sufficient?

```ruby
safe_music = mksafe(dynamic_music)
promos = mksafe(request.dynamic(dynamic_promo))
weighted = random(weights=[2, 1], [safe_music, promos])
```
`weighted` will play silence if `promos` or `safe_music` fails because of `mksafe`.

**Reply by tdardly** [reply]
No, probably don't need `request.dynamic` - that was one of the things I tried to say to liquidsoap "Stop running ahead and reading 6 tracks a minute!". As mentioned above, I now know that it was doing that because of the additional input/output in the script. Would love to know why this happens.

> weighted will play silence if promos or safe_music fails because of mksafe.

Not ideal, I will look again and fallbacks when I have found the best way to let the two "nodes" operate completely independently.

**Reply by vitoyucepi** [reply]
Sorry, I forgot about the main problem.
> Stop running ahead and reading 6 tracks a minute

Could you please post some logs produced by the server. It would also be helpful if you provide information about your setup.
1. Operating system.
1. Liquidsoap version.
1. Liquidsoap installation method.

**Reply by tdardly** [reply]
I'll look at enabling debug logging and get some logs sorted

In the meantime:
1. Alpine, running on Ubuntu
2. 2.2.3
3. docker-compose with savonet/liquidsoap-alpine

**Reply by vitoyucepi** [reply]
The default log level is sufficient for now.
To enable debug logging use `settings.log.level := 5`.

**Reply by tdardly** [reply]
Should the original config at the top actually work (I mean, with the second input/output)? If so, I'll fire up another docker, but I'm wondering if it doesn't by design (duplicate naming or whatever)...

**Reply by vitoyucepi** [reply]
Try to run this script
```ruby
# Configuration Variables
MUSIC_DIR = "/data/music"
PROMO_DIR = "/data/promo"
MUSIC_PLAYLIST = "#{MUSIC_DIR}/pl.m3u"
ICECAST_HOST = "my.icecast.server"
ICECAST_PORT = 80
ICECAST_PASS = "hackme"
MUSIC_MOUNT = "/128.mp3"

music = playlist(MUSIC_PLAYLIST)
promos = playlist(PROMO_DIR)
weighted = random(weights=[2, 1], [music, promos])

# Output to Icecast as MP3
output.icecast(
  %mp3,
  weighted,
  host = ICECAST_HOST,
  port = ICECAST_PORT,
  password = ICECAST_PASS,
  mount = MUSIC_MOUNT,
  fallible = true,
)
```

**Reply by tdardly** [reply]
Yes that will definitely work - as I say, it's the second input/output that causes the issue.

**Reply by vitoyucepi** [reply]
Hi @toots,
There's definitely a problem here.
When I use a chain that looks like
```mermaid
flowchart LR;
    playlist-->output1[output.icecast];
    output1[output.icecast]-->icecast1((icecast/a));
    icecast1((icecast/a))-->input.http;
    input.http-->output2[output.icecast];
    output2[output.icecast]-->icecast2((icecast/b));
```
then the playlist will consume inputs as fast as possible.

@tdardly,
I've found that it's possible to create a new clock that decouples the outputs.
```ruby
music = playlist("/tmp/test/music")

output.icecast(
  %mp3,
  music,
  host = "icecast",
  mount = "/a",
  fallible = true,
)

input_stream = clock(input.http("http://icecast:8000/a"))
                                                                                                         
output.icecast(                                                                                          
  %mp3,                                                                                                  
  input_stream,                                                                                          
  host = "icecast",                                                                                      
  mount = "/b",                                                                                          
  fallible = true,                                                                                       
) 
```

But you can use liquidsoap as a middleman and redirect the main stream to [`input.harbor`](https://www.liquidsoap.info/doc-2.2.4/reference.html#input.harbor).
This allows you to encode multiple streams at once without an additional Icecast jump.
```ruby
music = playlist("/tmp/test/music")
live = input.harbor()

stream = fallback([live, music], track_sensitive=false)

output.icecast(
  %mp3,
  stream,
  host = "icecast",
  mount = "/a",
  fallible = true,
)

output.icecast(
  %ogg,
  stream,
  host = "icecast",
  mount = "/b",
  fallible = true,
)
```

**Reply by tdardly** [reply]
Relieved you could recreate the issue @vitoyucepi - thank you.

The clock solution is good, I wondered whether a clock might be the answer here. 

I keep meaning to look into harbor, too. That would make it easy to use an uncompressed source as well, I guess, to minimise quality loss when re-encoding to two formats?

**Reply by vitoyucepi** [reply]
> That would make it easy to use an uncompressed source as well, I guess, to minimise quality loss when re-encoding to two formats?

You can use any supported lossy codec with reasonable settings, the result quality after encoding to *mp3-128k* will be the same for *flac-700k* and *mp3-192k*. Opus will be fine too.

**Reply by tdardly** [reply]
Can you confirm what you mean there - are you saying there would be no perceivable difference between flac->aac and mp3->aac ? Or there would be?

**Reply by vitoyucepi** [reply]
I think the quality of the result will mostly be defined by the limits of the output codecs, so it's not necessary to send an uncompressed stream to liquidsoap's `input.harbor`.

If you plan to broadcast *aac-64k* and *mp3-128k* to icecast, it will be fine to use *mp3-192k*, *aac-160k* or *opus-160k* to stream to liquidsoap.
